nearcore/src/shard_tracker.rs

```diff
@@ -1,95 +1,69 @@
-    actual_tracked_shards: HashSet<ShardId>,
+    tracked_accounts: Vec<AccountId>,
+    /// Stores tracking information by epoch
+    tracking_shards: RwLock<HashMap<EpochId, ShardBitMask>>,
+    /// Whether to track all shards
+    track_all_shards: bool,
     /// Epoch manager that for given block hash computes the epoch id.
     epoch_manager: Arc<RwLock<EpochManager>>,
-    /// Current shard layout
-    /// TODO: ShardTracker does not work if shard_layout is changed,
-    ///       fix this when https://github.com/near/nearcore/pull/4668 is merged
-    shard_layout: ShardLayout,
 }
 
 impl ShardTracker {
     pub fn new(
         accounts: Vec<AccountId>,
-        shards: Vec<ShardId>,
+        track_all_shards: bool,
         epoch_manager: Arc<RwLock<EpochManager>>,
     ) -> Self {
-        let shard_layout = {
-            let mut epoch_manager = epoch_manager.write().expect(POISONED_LOCK_ERR);
-            epoch_manager.get_shard_layout(&EpochId::default()).unwrap().clone()
-        };
-        let tracked_accounts = accounts.into_iter().fold(HashMap::new(), |mut acc, x| {
-            let shard_id = account_id_to_shard_id(&x, &shard_layout);
-            acc.entry(shard_id).or_insert_with(HashSet::new).insert(x);
-            acc
-        });
-        let tracked_shards: HashSet<_> = shards.into_iter().collect();
-        let mut actual_tracked_shards = tracked_shards.clone();
-        for (shard_id, _) in tracked_accounts.iter() {
-            actual_tracked_shards.insert(*shard_id);
-        }
-        info!(target: "runtime", "Tracking shards: {:?}", actual_tracked_shards);
         ShardTracker {
-            tracked_accounts,
-            tracked_shards,
-            actual_tracked_shards,
+            tracked_accounts: accounts,
+            tracking_shards: RwLock::new(HashMap::new()),
+            track_all_shards,
             epoch_manager,
-            shard_layout,
         }
     }
 
-    fn track_account(&mut self, account_id: &AccountId) {
-        let shard_id = account_id_to_shard_id(account_id, &self.shard_layout);
-        self.tracked_accounts
-            .entry(shard_id)
-            .or_insert_with(HashSet::new)
-            .insert(account_id.clone());
-        self.actual_tracked_shards.insert(shard_id);
-    }
-
-    /// Track a list of accounts. The tracking will take effect immediately because
-    /// even if we want to start tracking the accounts in the next epoch, it cannot harm
-    /// us to start tracking them earlier.
-    #[allow(unused)]
-    pub fn track_accounts(&mut self, account_ids: &[AccountId]) {
-        for account_id in account_ids.iter() {
-            self.track_account(account_id);
+    fn tracks_shard_at_epoch(
+        &self,
+        shard_id: ShardId,
+        epoch_id: &EpochId,
+    ) -> Result<bool, EpochError> {
+        let mut tracking_shards = self.tracking_shards.write().expect(POISONED_LOCK_ERR);
+        if !tracking_shards.contains_key(epoch_id) {
```

Comment:
This looks like a standard once_cell-like caching idiom, so we might want to not write the `RwLock` at the call-site, but rather encapsulate it behind some API.

```
struct AppendOnlyMap {}

impl AppendOnlyMap {
    fn get_or_insert(&/*no mut*/ self, key: K, value: FnOnce() -> V) -> &V
}
```

We have something like that in the SDK: https://github.com/near/near-sdk-rs/blob/fc56a7913c1538e42a438d7373add6d620e5a462/near-sdk/src/utils/stable_map.rs

Also, you might find reading the docs/implementaton of OnceCell useful: https://docs.rs/once_cell/1.8.0/once_cell/. It should give a good intuition about making caches in Rust work.

Reply:
I moved the RWLock inside the API, but not sure if that is what you had in mind @matklad let me know!

In general, I think there are many cases in our code that violates interior mutability and causes quite some inconvenience. For example, `EpochManager` https://github.com/near/nearcore/blob/master/chain/epoch_manager/src/lib.rs#L62 stores caches for some epoch info. Mutability leaked from inside `SizedCache` to outside `EpochManager`, so the callers of `EpochManager` have to maintain a RwLock (https://github.com/near/nearcore/blob/master/nearcore/src/runtime/mod.rs#L1771) on `EpochManager` for any access to it. Same things happen for `ChainStore` or almost any other cache like struct because they all use `SizedCache`.

Reply:
I agree. We have some quite substandard code there :(